lib/bolt/shell.rb

```diff
@@ -0,0 +1,85 @@
+# frozen_string_literal: true
+
+module Bolt
+  class Shell
+    attr_reader :target, :conn, :logger
+
+    def initialize(target, conn)
+      @target = target
+      @conn = conn
+      @logger = Logging.logger[@target.safe_name]
+    end
+
+    def run_command(*_args)
+      raise NotImplementedError, "run_command() must be implemented by the shell class"
+    end
+
+    def upload(*_args)
+      raise NotImplementedError, "upload() must be implemented by the shell class"
+    end
+
+    def run_script(*_args)
+      raise NotImplementedError, "run_script() must be implemented by the shell class"
+    end
+
+    def run_task(*_args)
+      raise NotImplementedError, "run_task() must be implemented by the shell class"
+    end
+
+    def provided_features
+      []
+    end
+
+    def default_input_method(_executable)
+      'both'
+    end
+
```

Comment:
I don't think these need to go in separate files, but I do feel like this class is really doing 2 separate things: defining the API for shell classes, and defining common helpers that shells can use. Might be worth at least a comment at this line distinguishing between the two?